### PR TITLE
[BPK-4298] Programatically define menus

### PR DESCRIPTION
**Pull request by georgegillams**
+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/master/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] `UNRELEASED.md`
+ [ ] `README.md`
+ [ ] Tests
+ [ ] [Screenshotting code](https://github.com/Skyscanner/backpack-ios/blob/master/Example/Backpack%20Screenshot/Screenshots.swift)
+ [ ] Adding a component? Remember to expose it in the [main `Backpack.h` header file](https://github.com/Skyscanner/backpack-ios/tree/master/Backpack/Backpack.h)
+ [ ] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: Example/Backpack/ViewControllers/BPKRootListTableViewController.swift [hunk 1]
```diff
@@ -20,22 +20,75 @@ import UIKit
 import Backpack
 
 class BPKRootListTableViewController: UITableViewController {
```
Comment (georgegillams) [inline Example/Backpack/ViewControllers/BPKRootListTableViewController.swift:22]: We should probably split this into two classes, one reponsible for app structure and one responsible for the settings

File: Example/Backpack/NavigationData.swift [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+protocol Presentable {
+    func makeViewController() -> UIViewController
+}
+
+@objc
+class NavigationData: NSObject {
```
Comment (k0nserv) [inline Example/Backpack/NavigationData.swift:24]: Why does this need to be `@objc` and inherit from `NSObject`?

File: Example/Backpack/NavigationData.swift [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    struct Item {
+        var name: String?
```
Comment (k0nserv) [inline Example/Backpack/NavigationData.swift:26]: Why `String?` All items have a name, no?

File: Example/Backpack/NavigationData.swift [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+ * limitations under the License.
+ */
+
+protocol Presentable {
+    func makeViewController() -> UIViewController
+}
+
+@objc
+class NavigationData: NSObject {
+    struct Item {
```
Comment (k0nserv) [inline Example/Backpack/NavigationData.swift:25]: Should we call this `Story` to align with the terminiology in the web world?

File: Example/Backpack/NavigationData.swift [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        var name: String?
+        var result: Presentable
+    }
+
+    struct Section {
+        var name: String?
+        var items: [Item]?
```
Comment (k0nserv) [inline Example/Backpack/NavigationData.swift:32]: Should both these be optional if so why?

File: Example/Backpack/NavigationData.swift [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    static var bottomSheetItem = Item(name: "Bottom sheet", result: AppStructure(sections: [ Section(name: "", items: [
+        Item(name: "With scroll view", result: Storyboard(storyboardName: "BottomSheet", storyboardIdentifier: "BottomSheetScrollableContentViewController")),
+        Item(name: "With bottom section", result: Storyboard(storyboardName: "BottomSheet", storyboardIdentifier: "BottomSheetScrollableContentViewController")),
+        Item(name: "Presenting another sheet", result: Storyboard(storyboardName: "BottomSheet", storyboardIdentifier: "BottomSheetScrollableContentViewController")),
+        Item(name: "Non-scrollable content", result: Storyboard(storyboardName: "BottomSheet", storyboardIdentifier: "BottomSheetScrollableContentViewController")),
+        Item(name: "Resizeable content height", result: Storyboard(storyboardName: "BottomSheet", storyboardIdentifier: "BottomSheetScrollableContentViewController"))
+    ])]))
+    static var buttonsItem = Item(name: "Buttons", result: AppStructure(sections: [Section(name: "", items: [
+        Item(name: "Primary", result: Storyboard(storyboardName: "Buttons", storyboardIdentifier: "ButtonViewController")),
```
Comment (k0nserv) [inline Example/Backpack/NavigationData.swift:81]: There's a lot of repition in these declarations, is that something we could improve?

File: Example/Backpack/NavigationData.swift [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+        Item(name: "With delegate", result: Storyboard(storyboardName: "Snackbar", storyboardIdentifier: "SnackbarViewController"))
+    ])]))
+    static var spinnersItem = Item(name: "Spinners", result: Storyboard(storyboardName: "Spinners", storyboardIdentifier: "SpinnersViewController"))
+    static var starItem = Item(name: "Star ratings", result: AppStructure(sections: [Section(name: "", items: [
+        Item(name: "Interactive", result: Storyboard(storyboardName: "StarRatings", storyboardIdentifier: "StarRatingsViewController")),
+        Item(name: "Docs", result: Storyboard(storyboardName: "StarRatings", storyboardIdentifier: "StarRatingsViewController"))
+    ])]))
+    static var switchesItem = Item(name: "Switches", result: Storyboard(storyboardName: "Switches", storyboardIdentifier: "SwitchesViewController"))
+    static var tabItem = Item(name: "Tab bar controller", result: Storyboard(storyboardName: "TabBarControllers", storyboardIdentifier: "TabBarControllerViewController"))
+    static var tappableItem = Item(name: "Tappable link labels", result: AppStructure(sections: [Section(name: "", items: [
```
Comment (k0nserv) [inline Example/Backpack/NavigationData.swift:194]: If the `name` of  a `Section` is optional isn't this one of the cases where we'd want to use that over an empty string?

**Review by k0nserv — COMMENTED** [review]
Looking really good just a few comments from me 👍🏼 

I'm curious if we can do something with [Function Builders](https://www.swiftbysundell.com/articles/deep-dive-into-swift-function-builders/) to make it even nicer to write these?

**Pull request closed without merging**